src/bit_machine/mod.rs

```diff
@@ -193,11 +197,14 @@ impl BitMachine {
 
     /// Add a read frame with some given value in it, as input to the
     /// program
-    pub fn input(&mut self, input: &Value) {
-        // FIXME typecheck this
+    pub fn input(&mut self, input: Arc<Value>) -> Result<(), ExecutionError> {
+        if !input.clone().is_of_type(self.source_ty.as_ref()) {
```

Comment:
In 23bdebe5ef4f2c4dc622ba0bb1d2c783839d887b:

Please remove this `clone`. Also you shouldn't need to explicitly call `as_ref`. Just pass `&self.source_ty`.

Reply:
I overlooked this instance of `clone`. Thanks for pointing this out.